src/mpl/src/gpu/mpl_gpu_ze.c

```diff
@@ -225,6 +226,163 @@ int MPL_gpu_dev_affinity_to_env(int dev_count, char **dev_list, char **env)
     return ret;
 }
 
+int MPL_gpu_init_device_mappings(int max_devid, int max_subdevid)
+{
+    int mpl_err = MPL_SUCCESS;
+    int global_dev_count = max_devid + 1;
+    int global_subdev_count = 0;
+
+    /* If max_subdevid is 0, then all procs use tile 0 as root devices, so subdevices aren't
+     * needed. */
+    if (max_subdevid == 0) {
+        global_ze_device_count = global_dev_count;
+    } else {
+        /* We can still have the situation where all procs use non-zero tile as root devices, but
+         * this can't be detected unless we also reduce subdevice count. Thus, consider them as
+         * subdevices in the global_to_local_map even if they are all root devices. */
+        global_subdev_count = max_subdevid + 1;
+        global_ze_device_count = global_dev_count * (global_subdev_count + 1);
+    }
+
+    /* Initialize local_to_global_map */
+    local_to_global_map = MPL_malloc(local_ze_device_count * sizeof(int), MPL_MEM_OTHER);
+    if (local_to_global_map == NULL) {
+        mpl_err = MPL_ERR_GPU_NOMEM;
+        goto fn_fail;
+    }
+
+    for (int i = 0; i < local_ze_device_count; ++i) {
+        local_to_global_map[i] = -1;
+    }
+
+    /* Initialize global_to_local_map */
+    global_to_local_map = MPL_malloc(global_ze_device_count * sizeof(int), MPL_MEM_OTHER);
+    if (global_to_local_map == NULL) {
+        mpl_err = MPL_ERR_GPU_NOMEM;
+        goto fn_fail;
+    }
+
+    for (int i = 0; i < global_ze_device_count; ++i) {
+        global_to_local_map[i] = -1;
+    }
+
+    /* Parse ZE_AFFINITY_MASK to find device and subdevice visibility based on global device id. */
+    char *visible_devices = getenv("ZE_AFFINITY_MASK");
+    if (visible_devices) {
+        size_t len = strlen(visible_devices);
+
+        char *devices = MPL_malloc(len + 1, MPL_MEM_OTHER);
+        if (devices == NULL) {
+            mpl_err = MPL_ERR_GPU_NOMEM;
+            goto fn_fail;
+        }
+        char *free_ptr, *dev;
+
+        memcpy(devices, visible_devices, len + 1);
```

Comment:
An overall comment that it would be nice to not have to re-parse the affinity mask string in this function. Or at least not duplicate the parsing code between here and `get_max_dev_id_from_mask`. Would it be possible to define some data structure to store the affinity mask information for various functions that need it?

Reply:
Done; I have added `parse_affinity_mask` which stores the contents of ZE_AFFINITY_MASK into a more suitable format that can be reused between `get_max_dev_id` and `MPL_gpu_init_device_mappings`

Reply:
Thanks. Looks good to me. Will approve and merge pending tests.